src/dakota.py

```diff
@@ -104,41 +127,37 @@ def __init__(self, **kwargs):
         for key in kwargs:
             setattr(self, key, kwargs[key])
 
-    def write_input(self, infile, data=None, other_data=None):
+    def write_input(self, infile, driver_instance=None):
         """
         Write input file sections in standard order.
         If data is not None, its id is written to ``analysis_components``.
         The invoked Python method should then recover the original object
         using :meth:`fetch_data`.
         """
-        print 'RUNNING'
         with open(infile, 'w') as out:
-            for section in ('environment', 'method', 'model', 'variables',
-                            'interface', 'responses'):
+            for section in ('environment', 'method', 'model', 'variables', 'interface', 'responses'):
+                # Write the section and all its sub keywords
                 out.write('%s\n' % section)
                 for line in getattr(self, section):
                     out.write("\t%s\n" % line)
 
-                if section == 'interface' and data is not None:
+                # Write the driver instance id as analysis_components
+                # if driver instance is specified and store the reverence to this instance for later use
+                if section == 'interface' and driver_instance is not None:
```

Comment:
It can be that you still want to use an external model (external application) as a driver. This is the whole reason for this checks here. I do not know if we want to still keep this in carolina solution or make it Python only. As it is right now the callback function is not ready to run anything but the driver anyway. If you say you would rather see a clean, fit for purpose solution I will clean all this up now.